enterprise/server/util/oci/oci.go

```diff
@@ -110,3 +121,107 @@ func (c Credentials) String() string {
 func (c Credentials) Equals(o Credentials) bool {
 	return c.Username == o.Username && c.Password == o.Password
 }
+
+func Resolve(ctx context.Context, imageName string, platform *rgpb.Platform, credentials Credentials) (v1.Image, error) {
+	imageRef, err := ctrname.ParseReference(imageName)
+	if err != nil {
+		return nil, status.InvalidArgumentErrorf("invalid image %q", imageName)
+	}
+
+	remoteOpts := []remote.Option{remote.WithContext(ctx)}
+	if !credentials.IsEmpty() {
+		remoteOpts = append(remoteOpts, remote.WithAuth(&authn.Basic{
+			Username: credentials.Username,
+			Password: credentials.Password,
+		}))
+	}
+
+	remoteDesc, err := remote.Get(imageRef, remoteOpts...)
+	if err != nil {
+		if t, ok := err.(*transport.Error); ok && t.StatusCode == http.StatusUnauthorized {
+			return nil, status.PermissionDeniedErrorf("could not retrieve image manifest: %s", err)
+		}
+		return nil, status.UnavailableErrorf("could not retrieve manifest from remote: %s", err)
+	}
+
+	switch remoteDesc.MediaType {
+	// This is an "image index", a meta-manifest that contains a list of
+	// {platform props, manifest hash} properties to allow client to decide
+	// which manifest they want to use based on platform.
+	case types.OCIImageIndex, types.DockerManifestList:
+		imgIdx, err := remoteDesc.ImageIndex()
+		if err != nil {
+			return nil, status.UnknownErrorf("could not get image index from descriptor: %s", err)
+		}
+		imgs, err := partial.FindImages(imgIdx, match.Platforms(v1.Platform{
+			Architecture: platform.GetArch(),
+			OS:           platform.GetOs(),
+			Variant:      platform.GetVariant(),
+		}))
+		if err != nil {
+			return nil, status.UnavailableErrorf("could not search image index: %s", err)
+		}
+		if len(imgs) == 0 {
+			return nil, status.NotFoundErrorf("could not find suitable image in image index")
+		}
+		if len(imgs) > 1 {
+			return nil, status.NotFoundErrorf("found multiple matching images in image index")
+		}
+		return imgs[0], nil
+	case types.OCIManifestSchema1, types.DockerManifestSchema2:
+		img, err := remoteDesc.Image()
+		if err != nil {
+			return nil, status.UnknownErrorf("could not get image from descriptor: %s", err)
+		}
+		return img, nil
+	default:
+		return nil, status.UnknownErrorf("descriptor has unknown media type %q", remoteDesc.MediaType)
+	}
+}
+
+func Authorized(ctx context.Context, image v1.Image, imageName string, credentials Credentials) error {
```

Comment:
nit: call this `authorizeLayers` ?

Reply:
Removed this function per above comment.